src/coldfront_plugin_api/urls.py

```diff
@@ -10,16 +11,68 @@
 
 
 class AllocationViewSet(viewsets.ReadOnlyModelViewSet):
+    """
+    This viewset implements the API to Coldfront's allocation object
+    The API allows filtering allocations by any of Coldfront's allocation model attributes,
+    as well as by Allocation Attributes.
+
+    Filtering by allocation model attributes
+    (i.e by pk, start_date, or even by attributes that span relationships, like project PI email)
+    is possible by using the query parameters directly as Django query filters.
+
+    For example, to filter by the allocation's project PI email: "/api/allocations?project__pi__email=test@bu.edu"
+
+    Filtering using this method requires knowledge of the Allocation model schema and Django's filters
+    Documentation of Coldfront Allocation model: https://coldfront.readthedocs.io/en/latest/apidocs/allocations/
+    Documentation for Django's query syntax: https://docs.djangoproject.com/en/5.0/topics/db/queries/#lookups-that-span-relationships
+
+    To filter by allocation attribute (AA) (i.e Quota Limits CPU),
+    name the query as the AA (case insensitive), prefix with "attr_", and replace all space with "_"
+
+    For all filters, you can also filter with conditional OR,
+    by providing multiple values seperated by comma
+
+    For example, to find allocations with AA Quota Limits CPU equal to 2 OR 4:
+    "/api/allocations?attr_quota_limits_cpu=2,4"
+    """
+
     serializer_class = serializers.AllocationSerializer
     authentication_classes = auth.AUTHENTICATION_CLASSES
     permission_classes = [IsAdminUser]
 
     def get_queryset(self):
         queryset = Allocation.objects.filter(status__name="Active")
+        query_params = self.request.query_params
 
-        if self.request.query_params.get("all") == "true":
+        if query_params.get("all") == "true":
             queryset = Allocation.objects.all()
 
+        allocation_attr_prefix = "attr_"
+        query_split_char = ","
+
+        for query, val in query_params.items():
+            if query == "all":
+                continue
+            if query.startswith(allocation_attr_prefix):
+                attribute_query = query[len(allocation_attr_prefix) :].replace("_", " ")
+                val_list = val.split(query_split_char)
```

Comment:
I think all of these functions are offered by `QueryDict`

Have you tried [getlist()](https://docs.djangoproject.com/en/5.0/ref/request-response/#django.http.QueryDict.getlist)?

Reply:
I not sure which `QueryDict` functions you think could be applied here. `getlist()` returns the value of a specified key, but I'm not sure how it would be applied here.

Reply:
According to the documentation for getlist:

> Returns a list of the data with the requested key

Which means it would handle the splitting for us.

Reply:
`getlist()` is used in the case where the same query parameter is named multiple times in the query string, i.e `?a=1&a=2`, in which case `getlist('a')` would return the list `[1,2]`. 

In this if statement, I am checking for query parameters starting with a specific prefix (i.e `?attr_OpenStack GPU Quota=1,2`), and split the query string with using a specifc character (in this code, its `,`). I see that in both cases, the end goal is to get a list, indicating the values we want to use when filtering allocations by a specific attributes.

I guess instead of doing string splitting ourselves to get the query parameter, you could format the query url as `?attr_OpenStack GPU Quota=1&attr_OpenStack GPU Quota=2`)`, and use `getlist()` there, but I thought that looked too long.

Reply:
I see. That is still preferable than doing the splitting ourselves.

Reply:
And can you also add a few tests to verify what happens when you try to filter for a non existing attribute?

Reply:
No, a 404 means a specific requested object was not found. Filtering isn't a request for a specific object, and therefore either a 400 Bad Request or an empty list is fine.